### PR TITLE
match old fuser rem to eager

### PR DESCRIPTION
**Pull request by Krovatkin**

### CODE REVIEW

**Review by jjsjann123 — APPROVED** [review]
Have some comment on tests.

File: test/test_jit_fuser.py [hunk 1]
```diff
@@ -171,6 +171,21 @@ def f(x, y):
...
+    @unittest.skipIf(not RUN_CUDA, "No CUDA")
+    def test_remainder_cuda(self):
+        def cuda_rem(x, y):
+            return 1 + torch.remainder(x, y) - 1
+
+        a = torch.tensor([-5.3]).cuda()
+        b = torch.tensor([-2]).cuda()
+        inputs = [a, b]
+        ge = self.checkScript(cuda_rem, inputs)
+        a = torch.tensor([-0.8]).cuda()
```
[the remark this answers is not in the document]
Comment (ngimel) [reply test/test_jit_fuser.py:183]: Having a test with hardcoded values is ok (but we should check different combinations of negative/positive divisor/dividend). As Jie suggests, integer test would also be useful. My problem with jit tests is that one never knows if the kernel has run. Even if the fusion shows in the graph, it's still possible that we bailed out at the last moment. That's a general issue though, not specific to this one, so does not need to be addressed.

[the remark this answers is not in the document]
Comment (Krovatkin) [reply test/test_jit_fuser.py:183]: i'll add integer tests and use `torch.rand()`

[the remark this answers is not in the document]
Comment (Krovatkin) [reply test/test_jit_fuser.py:183]: @ngimel @jjsjann123 I just realized the fuser codegen doesn't support templates for integers, only floats and doubles.
Basically, whenever we use those templates and pass integer tensors we will run into an error like this:

```
RuntimeError: default_program(66): error: more than one instance of overloaded function "fabs" matches the argument list:
            function "fabs(double)"
            function "fabs(float)"
            argument types are: (int)

default_program(97): error: more than one instance of overloaded function "fabs" matches the argument list:
            function "fabs(double)"
            function "fabs(float)"
            argument types are: (int)
```

[the remark this answers is not in the document]
Comment (ngimel) [reply test/test_jit_fuser.py:183]: How does the generated code look? nvcc usually does not have a problem with fabs on int
https://godbolt.org/z/y22u7i

[the remark this answers is not in the document]
Comment (Krovatkin) [reply test/test_jit_fuser.py:183]: @ngimel we are assigning the result of `fabs` to an int variable.
```cpp
extern "C" __global__
void kernel_0(IndexType totalElements, const TensorInfo<int,1> t0, const TensorInfo<int,1> t1 ) {

  // check whether do vectorized load/store and allocate buffer
  bool flag_vec4 = true;
  int t0_buf[4];
  if(t0.sizes[0] % 4 != 0) flag_vec4 = false;
  if(((uint64_t) t0.data) % (4 * sizeof(int)) != 0) flag_vec4 = false;
  int t1_buf[4];
  if(t1.sizes[0] % 4 != 0) flag_vec4 = false;
  if(((uint64_t) t1.data) % (4 * sizeof(int)) != 0) flag_vec4 = false;

  if (flag_vec4) {
    for (IndexType linearIndex = 4 * (blockIdx.x * blockDim.x + threadIdx.x);
         linearIndex < totalElements;
         linearIndex += 4 * gridDim.x * blockDim.x) {
      // Convert `linearIndex` into an offset of tensor as it is:
      IndexType t0_offset = 0;
      IndexType t0_linearIndex = linearIndex;

      //printf("tensor t0 sizes[0] = %d, strides[0] = %d\n", t0.sizes[0],t0.strides[0]);
      size_t t0_dimIndex0 = t0_linearIndex ;
      t0_offset += t0_dimIndex0 ;
      IndexType t1_offset = 0;
      IndexType t1_linearIndex = linearIndex;

      //printf("tensor t1 sizes[0] = %d, strides[0] = %d\n", t1.sizes[0],t1.strides[0]);
      size_t t1_dimIndex0 = t1_linearIndex ;
      t1_offset += t1_dimIndex0 ;

      // load 4 at a time
      *(reinterpret_cast<float4*>(t0_buf)) = *(reinterpret_cast<float4*>(t0.data + t0_offset));

      #pragma unroll 4
      for (int i=0; i<4; i++) {
        // calculate the results
        int n0 = t0_buf[i];
        int64_t n1 = 1;
        int n2 = fabs(n0);
        int n3 = n2 + n1*n1;
        int n4 = (n3 - n1*n1);
        t1_buf[i] = n4;

      }
      // store 4 at a time
      *(reinterpret_cast<float4*>(t1.data + t1_offset)) = *(reinterpret_cast<float4*>(t1_buf));

    }
  } else {
    for (IndexType linearIndex = blockIdx.x * blockDim.x + threadIdx.x;
         linearIndex < totalElements;
         linearIndex += gridDim.x * blockDim.x) {
      // Convert `linearIndex` into an offset of tensor:
      IndexType t0_offset = 0;
      IndexType t0_linearIndex = linearIndex;

      //printf("tensor t0 sizes[0] = %d, strides[0] = %d\n", t0.sizes[0],t0.strides[0]);
      size_t t0_dimIndex0 = t0_linearIndex ;
      t0_offset += t0_dimIndex0 ;
      IndexType t1_offset = 0;
      IndexType t1_linearIndex = linearIndex;

      //printf("tensor t1 sizes[0] = %d, strides[0] = %d\n", t1.sizes[0],t1.strides[0]);
      size_t t1_dimIndex0 = t1_linearIndex ;
      t1_offset += t1_dimIndex0 ;

      // calculate the results
      int n0 = __ldg(&t0.data[t0_offset]);
      int64_t n1 = 1;
      int n2 = fabs(n0); // @#$
      int n3 = n2 + n1*n1;
      int n4 = (n3 - n1*n1);
      t1.data[t1_offset] = n4;

    }
  }
}
```

**Comment by ngimel** [comment]
godbolt is still fine with this https://godbolt.org/z/KRkudj. @jjsjann123 what's up with nvrtc? Are there some defines missing? @Krovatkin can you try std::abs instead of fabs?

**Comment by Krovatkin** [comment]
> godbolt is still fine with this https://godbolt.org/z/KRkudj. @jjsjann123 what's up with nvrtc? Are there some defines missing? @Krovatkin can you try std::abs instead of fabs?

@ngimel that probably not gonna to work since we are missing a `math` include.

**Comment by Krovatkin** [comment]
> > godbolt is still fine with this https://godbolt.org/z/KRkudj. @jjsjann123 what's up with nvrtc? Are there some defines missing? @Krovatkin can you try std::abs instead of fabs?
> 
> @ngimel that probably not gonna to work since we are missing a `math` include.

 error: namespace "std" has no member "abs"

I will try adding the math header in a template header in `resource_strings.h`

**Comment by ngimel** [comment]
Unfortunately, iirc, nvrtc cannot look for includes and requires pretty much everything to be pasted in compile unit. @jjsjann123 can you please look at this?

**Comment by Krovatkin** [comment]
The integer issue doesn't seem to be directly related to the original issue reported. I'm going to add rand tests and land this PR since at the very least it does fix `aten::remainder` to the same level of the breakiness as other ops.

**Comment by jjsjann123** [comment]
Sorry about nvrtc :sob: 
I agree with Nick that this seems totally not related to the bug this PR is targeting and we can go ahead and land it without that piece.

Alternatively, it might not be idea, but we can update `abs` to a ternary op: https://github.com/pytorch/pytorch/blob/1bc3e3cafb737a8b9c8fea1cdfb750e74e7b5186/torch/csrc/jit/codegen/fuser/codegen.cpp#L184
`{aten::abs, "${0} < 0 ? -${0} : ${0} "},`

**Pull request closed without merging**